### ISSUE TITLE
[feature request] Add panel with thumbnails as an alternative to bookmarks

### ISSUE THREAD

**Issue by deveee** [sumatrapdfreader/sumatrapdf#146]
As first, thanks for this lightweight pdf viewer. Previously I was using Okular, but it needs a lot of KDE's libraries as the dependencies.

In other pdf viewers you can usually change if you want to see bookmarks or thumbnails in left panel. It would be nice to see it in Sumatra.

**Comment by Jolar** [comment]
This would be a useful feature, especially for image archives/cbz/cbr files.

**Comment by Unknow0059** [comment]
It would be great if I had a thumbnail preview of all pages in the sidebar. 
I use that feature frequently in another, slower, PDF viewer, and this one doesn't have it. It would be convenient if it did.
However, not as a replacement for bookmarks.

**Comment by GitHubRulesOK** [comment]
Which is easier to use 
how much slower are you prepared to go 
when viewing multiple pages at once can take much longer to draw them all

If viewing just one thumbnail at the same time why bother waiting twice as long ?

![image](https://user-images.githubusercontent.com/12754913/77470797-2b64aa80-6e09-11ea-8936-7f1ec5277e62.png)

> I use that feature frequently in another, slower, PDF viewer, and this one doesn't have it.

 Yes that's perfectly right It would almost certainly be slower if it did.

**Comment by SumatraPeter** [comment]
> How long will you wait before switching off because of loading that 400 page book you want to see all the thumbnails for ?

I don't see what the big deal is. I have used comic book reader apps that even on an old Android tablet have no problem generating thumbnails fairly quickly using a background thread as soon as any CBx is opened. Unless one scrolls the thumbnails panel forward a large number of pages immediately after opening a comic, one can't even see the thumbnails being generated in the background. Moreover the page thumbnails can optionally be cached as well, so that subsequent viewing of the same file results in very fast thumbs display. Heck, even Word seems to do fairly well when it comes to generating thumbs for very large documents. Plus of course the better the hardware running Sumatra is, the less thumb generation/scrolling speeds should be a problem.

Bookmarks, as useful as they are, are obviously not a 1:1 replacement for thumbnails. So while the request might remain in limbo simply because no-one's interested in implementing it, I don't think slow speed should be used as an excuse to suggest that the feature won't be useful.

**Comment by GitHubRulesOK** [comment]
@SumatraPeter whilst many old file formats supported secondary / embedded thumbnails (possibly even multiple adaptive copies to interlace in turn) few tend to carry that extra baggage. PDF/CBX usually depend on scaled duplicate renderings of the full page
For the example given above on this machine it takes approx 10 seconds to load the "thumbnail" view and only 2 seconds to load any single page I select from favourites.

**Comment by SumatraPeter** [comment]
@GitHubRulesOK: I know fully well that PDF/CBx files don't contain embedded thumbs, which is why I talked about the process of generating them using a bg thread, caching etc.

Two points - First, I have no clue what sort of hardware "this machine" has, but even if I did, there's no point making a general statement about speed/performance when it is obviously gonna be a different experience for everyone depending on what they own.

Second, Adobe's bloated buggy software would be the last thing I'd ever use as an example of good programming (their Flash IMO ranks as one of the worst programs in history).

I'd love to see a competent dev actually write some fast tightly optimized code to handle this, and see how it performs on even semi-decent modern hardware. I honestly don't think the experience would be as terrible as you're making it out to be.

**Comment by Unknow0059** [comment]
> Yes that's perfectly right It would almost certainly be slower if it did.

I'm not implying that's the reason it is slower, it could be, but even if the thumbnails aren't on screen nor being generated it is generally slower, which is what I meant.

**Comment by GitHubRulesOK** [comment]
I am not misconstruing anything simply observing that in all likelihood the more graphics bloat you introduce over time the slower applications becomes, unless you force users to keep upgrading the hardware. Without doubt if early Concurrent DOS & Windows had been hardened and secured correctly to run on today's hardware we would not need to discuss slowdown nor would we need antivirus causing more interupts bloat etc.

This thread was initially a comparison with a nice lightweight Linux Pdf document reader then expanded to wider graphics formats such as CBx where visual Table Of Contents/Bookmarks/Thumbnails make more sense.

Here on windows okular is struggling to re-draw both the thumbnail and page view at the same time
My point is why waste time doing both
![image](https://user-images.githubusercontent.com/12754913/77529607-da47cb80-6e87-11ea-97e4-f563a56bbbdd.png)



Custom Bookmarks **are almost possible now** (undocumented features) but take time to develop and could be dropped

**Comment by Unknow0059** [comment]
I'll be on the lookout for the custom bookmarks feature.

**Comment by GitHubRulesOK** [comment]
@Unknow0059 
If you use a portable copy of [pre-release](https://www.sumatrapdfreader.org/prerelease.html) or daily and open a pdf that has no bookmarks (or even one that has) 

right click the new bookmarks option

You can play with the ToC editor

![image](https://user-images.githubusercontent.com/12754913/77673769-840b8300-6f82-11ea-9bdd-c2b03b26ce59.png)

Please raise any questions under a separate topic as this one is for thumbnails

**Comment by SumatraPeter** [comment]
> Here on windows okular is struggling to re-draw both the thumbnail and page view at the same time

A KDE app ported to Windows? Why does the slow behavior not surprise me? 🙄🙂 Meanwhile when I downloaded the same issue of Analog and opened it up in a portable copy of PDF-XChange Editor the thumbs panel got populated pretty fast using multiple background threads (so the main document UI and rendering didn't get locked up) and the data got cached as well. Not to mention one can tweak all these settings too:

![PDF-XChange Editor Performance Options](https://user-images.githubusercontent.com/15380332/77674616-e0d26200-6fb1-11ea-8e37-e9db31625ac2.png)

> My point is why waste time doing both nothing is being gained and user is only suffering a loss

I whole-heartedly disagree, since your argument presupposes that the process is going to be excruciatingly slow, _which would just not be the case_ with a well-coded app on a machine with even semi-decent hardware (I didn't exactly test PDF-XChange on a beast of a PC here - far from it!).

Not to start a fight, but frankly I don't understand why you're pushing this specious 'slow speed' argument so hard and advocating against this feature? The fact is that many apps _do_ implement this and many folks _do_ find it useful. Now, this issue might remain open purely because no-one's willing to do the actual work of writing and testing all the necessary code (yet), but why queer the pitch and repeatedly suggest to any dev reading this that the feature is worthless and not worth implementing at all?

**Comment by GitHubRulesOK** [comment]
@SumatraPeter 
I agree it is **very well implemented in Tracker products** and have used thumbnails in Word and hundreds of other graphics editors to increase productivity. Potentially benefiting thousands of times a year for 4 decades But when that feature impacts my usage (most noticeably in reviewing or reading apps) it is one of the first features to be switched off. If a programmer with graphics programming skills wishes to add thumbnail support I am not against it but hope they also work on other graphic improvements first

**Comment by SumatraPeter** [comment]
> I agree it is **very well implemented in Tracker products**

Glad that you've finally seen that it _can_ be done well in a PDF app too.

> But when that feature impacts my usage (most noticeably in reviewing or reading apps) it is one of the first features to be switched off.

Nobody ever suggested for a second that it would be foisted onto everyone with no means to be turned off (which is why your opposition to it was a bit confusing - I mean, let it be implemented whenever possible even if _you_ don't want it or find it slow!). Every app I've seen to date that offers a thumbs view/panel makes it optional. If implemented as envisaged, users will be able to show/hide it just like they can with Bookmarks and Favorites at present.

> hope they also work on other graphic improvements first

There's lots to be done so let's leave it to the dev(s) to prioritize.

P.S. After this recent hullabaloo I fully expect this issue to go back into hibernation for another 5 years or so till someone resurrects it again! 😆

**Comment by Unknow0059** [comment]
I downloaded Okular from the Microsoft Store just to try it out. The thumbnails don't lag down the experience.
I hope that this issue (or rather, the new one it's been moved to) does not go into hibernation.
By the way, thanks for showing me how to use custom bookmarks.

**Comment by PolGZ** [comment]
Well, here is another one that wishes this was an option. Maybe not a default one, and only users that want it can enable it knowing it might slow down the process... but I've been opening pdfs for ages in different linux apps and I never found them anything but lighting fast. Now, that I work I must use one of the company's windows computer, searched for an open-source PDF viewer app to just run away from that bloated sh*t from Adobe... but I can't really feel confortable without the thumbnails. When you have a few "medium/large" documents for many hours changing from one page to another, checking things, going back and forth... it does nothing buf make way faster the process of dealing with the document.

I think this is a productivity/confort/ease of use option that should be available. I would see as something good if it is not enabled by default or whatever, for those who just want it to open the quickest possible. In my case if I just wanted to give a quick look to some pdf on a web or mail, I woundn't even download it and just open it straight on the browser. Buf for those documents you deal with at work every week, that have a few diagrams/pictures, some tables... but more than just 5 or 6 pages, it is sooo much quicker to have the help of a thumbnail preview on the side...

I mean, in my case, it is making me stay, for now, with damn Adobe... which I hate with all my heart. So, I think this sould be concidered by the devs, if possible.

Thank you very much!

**Comment by GitHubRulesOK** [comment]
@PolGZ 
you may note from the discussion above that there are lightweight alternative options to adobe such as okular or the advocated Tracker eXchange products until and IF thumbnails were added to SumatraPDF.

**Comment by PolGZ** [comment]
@GitHubRulesOK I did search for tracker exchange, but not only that is not open source software, but also I was welcomed with an 40/50$ pricincing license. So inmediately closed that. 

And as for okular I could only find the windows store way to install, wich on personal computers I would refuse to use. What is the point on avoinding windows telemetry, using open source apps and having to install something trough my microsoft account on their private hosted app-service. So I discarted that as well (although, I would do it at work).

But due to your comment, I read a bit more thoroughly, and I saw it could be installed via chocolatey as well. I may try that. Thanks.

(Anyway, I wanted to leave that comment to kindly ask for that option. And I take this opportunity again to thank the team for their work)

**Comment by GitHubRulesOK** [comment]
@PolGZ 
Tracker will add watermarks to extended features and many users still rely on the older less commercial freemium "view" version
You could also look at testing out https://binary-factory.kde.org/job/Okular_Nightly_win64/

**Comment by Tback1** [comment]
This feature is most useful in working with TotalCMD to quickly preview whether a PDF is complete or not

**Comment by GitHubRulesOK** [comment]
@Tback1 
PDF preview in Windows 10 is now so good that even I, don't use SumatraPDF or Tracker Previews (preferring them both to be portable rather than installed with previews active) The added bonus of using New Edge as default PDF annotator is MS have restored the most basic extras of Read Aloud and Comment Inking (but do I use SumatraPDF for text comments and Tracker for anything more.

**Comment by avelican** [comment]
Foxit has this feature. The thumbnails render almost instantaneously while your scroll the page preview pane. As an aside, it renders entire pages (in the regular document view) 2-3x faster as well.

I have settled on Sumatra for now because the scrolling is much, much smoother, but finding pages is much slower and less pleasant for those two reasons.

**Comment by Tback1** [comment]
> @Tback1 PDF preview in Windows 10 is now so good that even I, don't use SumatraPDF or Tracker Previews (preferring them both to be portable rather than installed with previews active) The added bonus of using New Edge as default PDF annotator is MS have restored the most basic extras of Read Aloud and Comment Inking (but do I use SumatraPDF for text comments and Tracker for anything more.

I agree your opinion for the preview in Windows 10,but I ask the privew not for prievew ,but for check does the PDF is completed. but after years , I looking another way use a pdf plugin to check the pages instead for the privew

**Comment by regs01** [comment]
There is already previews generation on Home tab. So functionality is there. Shouldn't it simplify making sidebar with previews of all pages?

**Comment by BrilliantTrees** [comment]
+One 

Having downloaded the programme today for testing I like the look of Sumatra for Windows it has the speed and simplicity of MuPDF which I use on Android but I agree the inclusion of an option to toggle thumb nails of open PDF's on/off would be really useful Want to ditch FoxIt and have been testing Sumatra and Okular today and like the fact that Ockular has this feature as standard. 

Would appreciate this being added to a development roadmap as the request seems to have been around for a good few years now

**Comment by GitHubRulesOK** [comment]
@kjk 
closure ?

**Comment by kjk** [comment]
technically it's asking for having this also in a left sidebar, which would be in addition to this

**Comment by GitHubRulesOK** [comment]
@kjk well yes it requested  at the side and others want it on the right 
but also related is https://github.com/sumatrapdfreader/sumatrapdf/issues/776

**Comment by GitHubRulesOK** [comment]
@kjk is this in effect done "panel with thumbnails as an alternative to bookmarks" 
middle solves the desire of both left and right users. see the OP mentioned okular aproach which is in ADDITION to bookmarks

<img width="1225" height="786" alt="Image" src="https://github.com/user-attachments/assets/6679ff69-8309-4a13-b0f7-cff3875ab695" />

<img width="1025" height="628" alt="Image" src="https://github.com/user-attachments/assets/03f2a724-f948-4444-b938-c61134409c33" />
